File: Raymond uber/eventListener.java

import java.io.*;
import java.util.*;
/*
设计一个EventListener要求实现register unregister 和 post 方法
*/
class Event {
  
  char id;
  public Event(char ID) {
    this.id = ID;
  }
  
  public void doEvent(String data) {
    System.out.println(id + data);
  }

  
  @Override
  public int hashCode() {
    return String.valueOf(id).hashCode();
  }
  
  @Override 
  public boolean equals(Object other) { //very important
    if(other instanceof Event) {
      return id == ((Event)other).id;
    }
    return false;
    
  }
}

class EventListener {
  Map<String,HashSet<Event>> map;
  
  public EventListener() {
    map = new HashMap<>();
  }
  
  public void register(String name , Event e) {
    if(!map.containsKey(name)) {
      map.put(name, new HashSet<Event>()); 
    }
    map.get(name).add(e);
  }
  
  public void unregister(String name, Event e) {
    if(!map.containsKey(name)) return;
    map.get(name).remove(e);
  }
  
  public void post(String name, String data) {
    if(!map.containsKey(name)) return;
    for(Event e : map.get(name)) {
      e.doEvent(data);
    }
  }
  
  
}



public class Solution {
  public static void main(String[] args) {
    EventListener Listener = new EventListener();
    Listener.register("div",new Event('A'));
    Listener.register("div", new Event('B'));
    //Listener.post("div", "hello");
    Listener.unregister("div", new Event('A'));
    Listener.post("div", "hello");
    Event e1 = new Event('A');
    Event e2 = new Event('A');
    System.out.println(e1.equals(e2));
    HashMap<Event,Integer> map = new HashMap<>();
    map.put(e1,2);
    System.out.println(e1.hashCode());
    System.out.println(e2.hashCode());
    System.out.println(e1.equals(e2));
  }
  
  
}

File: Raymond uber/eventListener.cpp
/*
设计一个EventListener要求实现register unregister 和 post 方法
*/
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/*class Event{
private:
	string Data;
public:
	char ID;
	Event(char id){
		ID=id;
	}
	void doEvent(string data){
		Data=data;
		switch(ID){
			case 'A': cout << Data << " functionA " << endl;
					   break;
			case 'B': cout << Data << " functionB " << endl;
				       break;
			default: cout << "none" << endl;
				 	  break;
		}

	}
};

inline bool operator == (Event const a, Event const b){return a.ID==b.ID;}
class EventListener {
struct hash{
	size_t operator() (const Event e) const{
		int id=e.ID-'0';
		return id;
	}
};
private:
	unordered_map< string, unordered_set<Event, hash > >table;
public:

};
int main()
{
	
	EventListener Listener;
	//Listener.register("first", Event('A'));
	

	return 0;
}*/
class Event{
public:
	char ID;
	Event(char id){
		ID=id;
	}
	void doEvent(string data) const{
		switch (ID){
			case'A': cout << data << " with functionA" << endl;
					break;
			case'B': cout << data << " with functionB" << endl;
					break;
			default: cout << "none" << endl;
					break;
		}
	}

};
inline bool operator == (Event const & a, Event const & b){return a.ID==b.ID;}
class EventListener {
struct hash{
	size_t operator() (Event const & e)const {
		int id=e.ID-'0';
		return id;
	}
};
private:
	unordered_map< string, unordered_set<Event, hash > >table;
public:
	void Register(string name, Event e){
		if(!table.count(name))table[name]={e};
		else table[name].insert(e);
	}
	void Unregister(string name, Event e){
		if(!table.count(name))return;
		table[name].erase(e);
	}
	void post(string name, string data){
		if(!table.count(name))return;
		for(auto iter=table[name].begin(); iter!=table[name].end(); iter++){
			iter->doEvent(data);
		}
	}
};
int main()
{
	
	EventListener Listener;
	Listener.Register("div", Event('A'));
	Listener.Register("div", Event('B'));
	Listener.post("div", "hello");
	

	return 0;
}